File: src/types/audio.ts
/**
 * Espelho TS de `AudioMedia` (Rust: `models/audio.rs`). Módulo Áudio, Camada 1.
 *
 * `relative_path` aponta para o WAV de análise (derivado determinístico);
 * `original_relative_path` é o original preservado (quando importado).
 */

export interface AudioMedia {
  id: string;
  occurrence_id: string;
  /** "importado" | "extraido" */
  kind: string;
  original_path: string | null;
  original_relative_path: string | null;
  relative_path: string;
  filename: string;
  sha256: string;
  original_sha256: string | null;
  source_video_sha256: string | null;
  size_bytes: number;
  duration_s: number | null;
  sample_rate: number | null;
  channels: number | null;
  codec: string | null;
  bitrate: number | null;
  raw_probe_json: string;
  warnings_json: string;
  created_at: string;
  updated_at: string;
}

/** Espelho de `AudioMarker` (Rust). Marcador temporal persistido no caso. */
export interface AudioMarker {
  id: string;
  occurrence_id: string;
  audio_sha256: string;
  t_seconds: number;
  label: string;
  created_at: string;
}

/**
 * Espelho de `AudioTranscriptSegment` (Rust). Segmento de degravação MANUAL
 * (trabalho do perito): trecho temporal com rótulo de locutor e texto.
 */
export interface AudioTranscriptSegment {
  id: string;
  occurrence_id: string;
  audio_sha256: string;
  idx: number;
  t_start: number;
  t_end: number | null;
  speaker: string;
  text: string;
  created_at: string;
}

/** Status da ferramenta whisper.cpp (Fase 2 — transcrição local). */
export interface WhisperStatus {
  available: boolean;
  path: string | null;
}

/** Candidato de transcrição (RASCUNHO de máquina) devolvido pelo whisper. */
export interface TranscriptCandidate {
  idx: number;
  t_start: number;
  t_end: number | null;
  speaker: string;
  text: string;
  /** Confiança média (0..1) da IA neste trecho; null se indisponível. */
  confidence: number | null;
}

// W12 (paridade Audacity) — Análise forense (espelho de `audio/analysis.rs`).

/** Medições objetivas de um áudio (pico/RMS/DC/clipping/crista). */
export interface AudioMeasurements {
  duration_s: number;
  sample_rate: number;
  channels: number;
  samples: number;
  peak_linear: number;
  peak_dbfs: number;
  rms_dbfs: number;
  crest_factor_db: number;
  dc_offset: number;
  dc_offset_pct: number;
  clipped_samples: number;
  clipped_runs: number;
  clipped_pct: number;
}

/** Espectro Welch (FFT janelada + média). */
export interface SpectrumResult {
  sample_rate: number;
  fft_size: number;
  window: string;
  freqs_hz: number[];
  mag_db: number[];
  peak_freq_hz: number;
  peak_db: number;
}

/** Curva ENF (Electric Network Frequency) + continuidade. */
export interface EnfResult {
  nominal_hz: number;
  window_s: number;
  step_s: number;
  times_s: number[];
  enf_hz: number[];
  mean_hz: number;
  std_hz: number;
  /** Maior salto frame-a-frame (Hz): descontinuidade = indício de edição. */
  max_jump_hz: number;
}

File: src-tauri/src/models/audio.rs
//! Modelos do módulo Áudio (Camada 1).

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Uma mídia de áudio registrada numa ocorrência. O `relative_path` aponta
/// para o WAV de análise (derivado determinístico); o original importado fica
/// preservado em `original_relative_path` (quando `kind == "importado"`).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioMedia {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    /// "importado" | "extraido".
    pub kind: String,
    pub original_path: Option<String>,
    pub original_relative_path: Option<String>,
    pub relative_path: String,
    pub filename: String,
    /// Hash do WAV de análise (chave de dedupe na ocorrência).
    pub sha256: String,
    pub original_sha256: Option<String>,
    pub source_video_sha256: Option<String>,
    pub size_bytes: u64,
    pub duration_s: Option<f64>,
    pub sample_rate: Option<u32>,
    pub channels: Option<u32>,
    pub codec: Option<String>,
    pub bitrate: Option<i64>,
    pub raw_probe_json: String,
    pub warnings_json: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Marcador temporal posto pelo perito no player de áudio.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioMarker {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    /// Hash do WAV de análise ao qual o marcador pertence.
    pub audio_sha256: String,
    pub t_seconds: f64,
    pub label: String,
    pub created_at: DateTime<Utc>,
}

/// Registro de um realce (auxílio de escuta) aplicado a um áudio. Não-destrutivo:
/// a saída é uma nova mídia; aqui fica o vínculo origem→saída + cadeia FFmpeg.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioEnhancement {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub source_audio_sha256: String,
    pub output_audio_sha256: String,
    pub filters_json: String,
    pub created_at: DateTime<Utc>,
}

/// Um segmento de degravação MANUAL (trabalho do perito): trecho temporal com
/// rótulo de locutor e texto. O tool não transcreve nem interpreta — só guarda.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioTranscriptSegment {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub audio_sha256: String,
    pub idx: i64,
    pub t_start: f64,
    pub t_end: Option<f64>,
    pub speaker: String,
    pub text: String,
    pub created_at: DateTime<Utc>,
}

/// Entrada de segmento vinda do front (sem id/custódia — gerados ao salvar).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TranscriptSegmentInput {
    pub idx: i64,
    pub t_start: f64,
    pub t_end: Option<f64>,
    #[serde(default)]
    pub speaker: String,
    #[serde(default)]
    pub text: String,
}
